src/app/ReadHandler.cpp

```diff
@@ -579,21 +579,35 @@ CHIP_ERROR ReadHandler::ProcessSubscribeRequest(System::PacketBufferHandle && aP
     return CHIP_NO_ERROR;
 }
 
+void ReadHandler::OnUnblockHoldReportCallback(System::Layer * apSystemLayer, void * apAppState)
+{
+    VerifyOrReturn(apAppState != nullptr);
+    ReadHandler * readHandler = static_cast<ReadHandler *>(apAppState);
+    ChipLogProgress(DataManagement, "Unblock report hold after min %d seconds", readHandler->mMinIntervalFloorSeconds);
+    readHandler->mHoldReport = false;
```

Comment:
We do need to call ScheduleRun() here since we may have accumulated dirty items in the global dirty set that need to get flushed now that the hold off has elapsed.

Reply:
We need a distinction between:
1. When the min-interval has elapsed, to _only_ send out a report if there is actual data to be sent (modulo other criteria that may need to be satisfied, e.g batching parameters for events)
2. When the max-interval has elapsed, to _definitely_ send out a report _even if_ there is no actual data to be sent.

This distinction I don't think is present today.

Reply:
Just updated this. thanks

Reply:
I'm still not quite getting how the logic changes you've done achieve the distinction between the two cases I've outlined above. Can you please explain?

Reply:
in OnUnblockHoldReportCallback which is triggered when min-interval is elapsed, I check if current read handler is dirty, then decide if engine send the report out. in OnRefreshSubscribeTimerSyncCallback which is triggerred when max interval is elapsed, report engine would run.
i have updated the setDirty in engine.cpp which would mark read handler as dirty when there is intersected value between subscribed path and marked dirty-path

Reply:
> OnRefreshSubscribeTimerSyncCallback which is triggerred when max interval is elapsed, report engine would run

How do you ensure that even if there is no dirtiness, that an empty report will be sent at the max interval?